core/src/main/resources/reference.conf

```diff
@@ -550,6 +566,183 @@ datastax-java-driver {
+    session {
+      enabled = [
+        # The throughput and latency percentiles of CQL requests (exposed as a Timer).
+        #
+        # This corresponds to the overall duration of the session.execute() call, including any
+        # retry.
+        // cql_requests,
+
+        # The number of CQL requests that timed out -- that is, the session.execute() call failed
+        # with a DriverTimeoutException (exposed as a Counter).
+        // cql_client_timeouts,
+      ]
+
+      # Extra configuration (for the metrics that need it)
+      cql_requests {
+        # The largest latency that we expect to record.
+        #
+        # This should be slightly higher than request.timeout (in theory, readings can't be higher
+        # than the timeout, but there might be a small overhead due to internal scheduling).
+        #
+        # This is used to scale internal data structures. If a higher recording is encountered at
+        # runtime, it is discarded and a warning is logged.
+        highest_latency = 3 seconds
+
+        # The number of significant decimal digits to which internal structures will maintain value
+        # resolution and separation (for example, 3 means that recordings up to 1 second will be
+        # recorded with a resolution of 1 millisecond or better).
+        #
+        # This must be between 0 and 5. If the value is out of range, it defaults to 3 and a warning
+        # is logged.
+        significant_digits = 3
+
+        # The interval at which percentile data is refreshed.
+        #
+        # The driver records latency data in a "live" histogram, and serves results from a cached
+        # snapshot. Each time the snapshot gets older than the interval, the two are switched. Note
+        # that this switch happens upon fetching the metrics, so if you never fetch the recording
+        # interval might grow higher (that shouldn't be an issue in a production environment because
+        # you would typically have a metrics reporter that exports to a monitoring tool at a regular
+        # interval).
+        #
+        # In practice, this means that if you set this to 5 minutes, you're looking at data from a
+        # 5-minute interval in the past, that is at most 5 minutes old. If you fetch the metrics at
+        # a faster pace, you will observe the same data for 5 minutes until the interval expires.
+        #
+        # Note that this does not apply to the total count and rates (those are updated in real
+        # time).
+        refresh_interval = 5 minutes
+      }
+    }
+    # The node-level metrics (all disabled by default).
+    node {
+      enabled = [
+        # The number of connections open to this node for regular requests (exposed as a
+        # Gauge<Integer>).
+        #
+        # This does not include the control connection (which uses at most one extra connection to a
+        # random node in the cluster).
+        // pooled_connection_count,
```

Comment:
Just brainstorming (maybe i'm over thinking it).  A side effect of the granularity of being able to turn individual metrics on/off is that there are a few downsides:

1. If you want to include all metrics, you have to extract them all from this file and put them into your application.conf.  It's a bit tedious to have to copy the individual ones, but It's tolerable.

2. If a new metric is added, the user has to know about it and add it to their application.conf.

One possible solution would be to make the configuration more coarse and allow enabling metrics by category, i.e.:

```
enabled = [
  pooled_connection_count,
  available_stream_count,
  errors,
  retries,
  ignores
]
```

I wouldn't expect someone to selectively enable/disable specific retries for example, either they'd enable all of them or disable all of them.

you could also categorize `pooled_connection_count` , `available_stream_count` under a `pool` group or something.

Reply:
> If a new metric is added, the user has to know about it and add it to their application.conf

Yes, but on the flip side if metrics are enabled by default, then you risk introducing new (potentially expensive) metrics in existing environment, without the user's notice. I've been pondering on this "enabled vs disabled by default" question since the beginning of this PR, I think I prefer being conservative. We'll advertise new metrics in release notes and the upgrade guide.

> If you want to include all metrics, you have to extract them all from this file and put them into your application.conf. It's a bit tedious

At least that encourages you to read all the descriptions 😉

Reply:
> At least that encourages you to read all the descriptions 😉

Haha fair enough!  There isn't much of a downside in educating what the metrics **actually mean** (especially when it comes to the timer-based ones) and it will be a one time task.   I think i'm 👍  on the current configuration.